### PR TITLE
fixes the linearGradient on dotnet

### PR DESCRIPTION
**Pull request by mortoray**
This needs to be tested on Windows: I tested only DotNet on OSX.  I'm including  reference photo (CoreGraphics output, which is the same as the DotNet output on OSX).

_Note: The fix isn't complete per-se, you can still create gradients that fail by making the gradient line not cross the fill region. It's not a normal use-case, and would take more effort to fix than I think is desired now._

Test project: https://github.com/mortoray/fork-sandbox/tree/master/Jan2018/LinearGrad

screen shot 2018-01-17 at 11 36 10

fixes https://github.com/fusetools/fuselibs-public/issues/936
fixes https://github.com/fusetools/fuselibs-public/issues/841

This PR contains:
- [x] Changelog
- [ ] ~Documentation~
- [ ] ~Tests~

### CODE REVIEW

**Comment by kusma** [comment]
Looks reasonable to me on Windows:
image

File: Source/Fuse.Drawing.Surface/DotNetSurface.uno [hunk 1]
```diff
@@ -674,6 +593,22 @@ namespace Fuse.Drawing
 			colors[numberOfColorPoints - 1] = endColor;
 			offsets[numberOfColorPoints - 1] = 1.0f;
 
+			var inVec = Vector.Normalize(inEnd - inStart);
+			var inLen = Vector.Length(inEnd - inStart);
+			//safety
+			if (inLen < 0.1f) {
```
Comment (kusma) [inline Source/Fuse.Drawing.Surface/DotNetSurface.uno]: I think this is a bit large of an epsilon here. Shouldn't the code work even if this is way smaller (like 1e-5 or something)?

Comment (mortoray) [reply Source/Fuse.Drawing.Surface/DotNetSurface.uno]: The code might start misbehaving if this gets too small as it alters the offset ranges too much.  0.1 is chosen as kind of a cut-off where anything below this you can't visually see a difference anyway (this is a measure of pixel-spread of the gradient).

**Comment by kusma** [comment]
I now get this when running manualtestingapp:
```
Error: System.Exception: Unexpected exception ---> System.OutOfMemoryException: Out of memory.
   at System.Drawing.Drawing2D.LinearGradientBrush..ctor(PointF point1, PointF point2, Color color1, Color color2)
   at Fuse.Drawing.DotNetHelpers.StrokePathLinearGradient(Graphics graphics, GraphicsPath path, LinearGradient lg, Single startX, Single startY, Single endX, Single endY, Single width, Single miterLimit, LineJoin lineJoin, LineCap lineCap) in C:\Users\erik\src\fuselibs\Source\Fuse.Drawing.Surface\DotNetSurface.uno:line 719
   at Fuse.Drawing.DotNetSurface.StrokePath(SurfacePath path, Stroke stroke) in C:\Users\erik\src\fuselibs\Source\Fuse.Drawing.Surface\DotNetSurface.uno:line 259
   at Fuse.Controls.Shape.Fuse.Drawing.ISurfaceDrawable.Draw(Surface surface) in C:\Users\erik\src\fuselibs\Source\Fuse.Controls.Primitives\Shapes\Shape.Surface.uno:line 105
   at Fuse.Drawing.Surface.DrawLocal(ISurfaceDrawable drawable) in C:\Users\erik\src\fuselibs\Source\Fuse.Drawing.Surface\Surface.uno:line 150
   at Fuse.Drawing.Surface.Draw(DrawContext dc, Element elm, ISurfaceDrawable drawable) in C:\Users\erik\src\fuselibs\Source\Fuse.Drawing.Surface\Surface.uno:line 135
   at Fuse.Controls.Shape.DrawVisual(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Controls.Primitives\Shapes\Shape.Visual.uno:line 22
   at Fuse.Controls.Control.OnDraw(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Controls\Control.uno:line 43
   at Fuse.Elements.Element.DrawWithChildren(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Batching.uno:line 113
   at Fuse.Elements.Element.CompositEffects(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Effects.uno:line 184
   at Fuse.Elements.Cache.DrawHeuristically(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Caching\Cache.uno:line 83
   at Fuse.Elements.Element.Composit(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Effects.uno:line 103
   at Fuse.Elements.Element.Draw(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Drawing.uno:line 109
   at Fuse.Elements.Element.DrawNonUnderlayChildren(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Batching.uno:line 71
   at Fuse.Elements.Element.DrawWithChildren(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Batching.uno:line 113
   at Fuse.Elements.Element.CompositEffects(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Effects.uno:line 184
   at Fuse.Elements.Cache.DrawHeuristically(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Caching\Cache.uno:line 83
   at Fuse.Elements.Element.Composit(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Effects.uno:line 103
   at Fuse.Elements.Element.Draw(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Drawing.uno:line 109
   at Fuse.Controls.Panel.Draw(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Controls.Panels\Panel.Freeze.uno:line 208
   at Fuse.Elements.Element.DrawNonUnderlayChildren(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Batching.uno:line 71
   at Fuse.Elements.Element.DrawWithChildren(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Batching.uno:line 113
   at Fuse.Elements.Element.CompositEffects(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Effects.uno:line 184
   at Fuse.Elements.Cache.DrawHeuristically(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Caching\Cache.uno:line 83
   at Fuse.Elements.Element.Composit(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Effects.uno:line 103
   at Fuse.Elements.Element.Draw(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Drawing.uno:line 109
   at Fuse.Controls.Panel.Draw(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Controls.Panels\Panel.Freeze.uno:line 208
   at Fuse.Elements.Element.DrawNonUnderlayChildren(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Batching.uno:line 71
   at Fuse.Elements.Element.DrawWithChildren(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Batching.uno:line 113
   at Fuse.Elements.Element.CompositEffects(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Effects.uno:line 184
   at Fuse.Elements.Cache.DrawHeuristically(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Caching\Cache.uno:line 83
   at Fuse.Elements.Element.Composit(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Effects.uno:line 103
   at Fuse.Elements.Element.Draw(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Drawing.uno:line 109
   at Fuse.Controls.Panel.Draw(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Controls.Panels\Panel.Freeze.uno:line 208
   at Fuse.Elements.Element.DrawNonUnderlayChildren(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Batching.uno:line 71
   at Fuse.Elements.Element.DrawWithChildren(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Batching.uno:line 113
   at Fuse.Elements.Element.CompositEffects(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Element.Effects.uno:line 184
   at Fuse.Elements.ElementAtlas.FillFramebuffer(DrawContext dc, Framebuffer fb, Boolean drawAll, Rect scissorRectInClipSpace) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Caching\ElementAtlas.uno:line 212
   at Fuse.Elements.ElementAtlas.PinAndValidateFramebuffer(DrawContext dc) in C:\Users\erik\src\fuselibs\Source\Fuse.Elements\Caching\ElementAtlas.uno:line 149
```

**Comment by mortoray** [comment]
Ah man, out-of-memory inside DotNet libraries isn't fun.  I'll see if it happens on OSX/DotNet.  It probably relates to the length of the gradient line -- if you can emit the gStart/gEnd prior to line 719 it might indicate the issue.

**Comment by kusma** [comment]
gStart: 0, 0
gEnd: 0, 0

:/

**Comment by kusma** [comment]
This confuses me... It seems like we should have exited earlier than this due to `if (bounds.IsEmpty)` at line 709...

**Comment by kusma** [comment]
Ah, the problem is that inVec gets zero-length.

**Comment by kusma** [comment]
So, StrokePathLinearGradient() gets called with startX=0, startY=0, endX=0, endY=0. That can't be very well-defined, can it?

**Comment by mortoray** [comment]
Completely zero length is weird for `inVec`... my epsilon check would catch low values, but I guess zero values still result in nonsense.

I can't reproduce this on DotNet/OSX, so Mono appears to check against this case.

**Comment by kusma** [comment]
The question is, what do we expect to happen here?

**Comment by mortoray** [comment]
No, if Start == End the gradient is bascially undefined. We could fix it however we want, I suggest setting start to 0,0 and End to 1,1.

This may happen if an element is currently a zero-sized (happens during scaling, but that'd be unusual for the test app)

**Comment by mortoray** [comment]
Sampling the middle of the gradient would be a lot more work in the surface implementation, since it's already in the gradient branch.

**Comment by kusma** [comment]
Sure, it just seems like a more well-behaved fallback than 0,0 <-> 1,1. That angle seems pretty arbitrary to me.

**Comment by kusma** [comment]
By the way `bounds: {X=-7,Y=-7,Width=14,Height=14}`, so it's not an invisible element. It's the gradient-points who are problematic, not the size of the element itself.

**Comment by mortoray** [comment]
The angle is arbitrary, but the default on DotNet, thus I assume it'll work on all sizes.  Plus, given the small size it'll likely sample the middle color, or the end color or any color, but since we say it's undefined we don't care which one.

**Comment by mortoray** [comment]
Aha, so the gradient points are too close together in that case (After my translation). I can probably improve my fix so that doesn't happen.

**Comment by kusma** [comment]
This cheesy fix gets things working for me again:
```patch
diff --git a/Source/Fuse.Drawing.Surface/DotNetSurface.uno b/Source/Fuse.Drawing.Surface/DotNetSurface.uno
index caba0a92fe..fefde896dc 100644
--- a/Source/Fuse.Drawing.Surface/DotNetSurface.uno
+++ b/Source/Fuse.Drawing.Surface/DotNetSurface.uno
@@ -597,7 +597,7 @@ namespace Fuse.Drawing
 			var inLen = Vector.Length(inEnd - inStart);
 			//safety
 			if (inLen < 0.1f) {
-				inVec = inEnd - inStart; //this affects only gStart/gEnd so okay to be incorrect in nonsense case
+				inVec = float2(1); //this affects only gStart/gEnd so okay to be incorrect in nonsense case
 				inLen = 1;
 			}
 
```
@mortoray: If you don't mind, I'd like to push this to master to un-break stuff. Is that OK?

**Comment by mortoray** [comment]
I'm fine with that fix. It's just an unusual corner-case anyway, and I said my code doens't handle them all correctly (like apparently DotNet doesn't either ;)

**Comment by mortoray** [comment]
If it works, and my test app renders the same, I'd say it's a suitable long-term fix -- until we find a legitimate use of a corner-case that produces the wrong gradient.

**Comment by kusma** [comment]
@mortoray: It looks correct as far as I can tell. I just went ahead and pushed out the fix, since things don't currently even start.
